### PR TITLE
expose oauth config in config

### PR DESCRIPTION
**Pull request by deads2k**
Exposes the oauth config.  I'll add integration tests once we're able to modify the actual config.

Here is an example of how it looks:
https://gist.github.com/deads2k/da5618636ce4bf83c0c2

@liggitt

### CODE REVIEW

File: pkg/cmd/server/api/types.go [hunk 1]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
 }
 
+type HTPasswdPasswordIdentityProvider struct {
+	api.TypeMeta
+
+	// File is a reference to your htpasswd file
+	File string
+}
+
+type XRemoteUserIdentityProvider struct {
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: RequestHeaderIdentityProvider?

File: pkg/cmd/server/api/types.go [hunk 2]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+	api.TypeMeta
+
+	// CAFile is a file with the trusted signer certs
+	CAFile string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: ClientCA for congruity?

File: pkg/cmd/server/api/types.go [hunk 3]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+	// SessionSecrets list the secret(s) to use to encrypt created sessions. Used by AuthRequestHandlerSession
+	SessionSecrets []string
+	// SessionMaxAgeSeconds specifies how long created sessions last. Used by AuthRequestHandlerSession
+	SessionMaxAgeSeconds int32
+	// SessionName is the cookie name used to store the session
+	SessionName string
+}
+
+type IdentityProviderUsage struct {
+	// ProviderScope is used to qualify the identities teturned by this provider
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: typo

File: pkg/cmd/server/api/types.go [hunk 4]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+	ProviderScope string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: ProviderName to match new field in Identity object?

File: pkg/cmd/server/api/v1/types.go [hunk 5]
```diff
@@ -147,9 +148,103 @@ type OAuthConfig struct {
 	// AssetPublicURL is used for building valid client redirect URLs for external access
 	AssetPublicURL string `json:"assetPublicURL"`
 
-	// all the handlers here
+	IdentityProviders []IdentityProvider `json:"identityProviders"`
+
+	GrantConfig GrantConfig `json:"grantConfig"`
+
+	SessionAuthenticationConfig *SessionAuthenticationConfig `json:"sessionAuthenticationConfig"`
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: SessionConfig

File: pkg/cmd/server/api/types.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: This takes the place of ProxyCA

File: pkg/cmd/server/origin/auth.go [hunk 6]
```diff
@@ -322,230 +251,213 @@ func (c *AuthConfig) getAuthorizeAuthenticationHandlers(mux cmdutil.Mux) (authen
...
 
-		state := external.DefaultState()
-		oauthHandler, err := external.NewExternalOAuthRedirector(oauthProvider, state, c.MasterPublicAddr+callbackPath, successHandler, errorHandler, identityMapper)
-		if err != nil {
-			return nil, fmt.Errorf("unexpected error: %v", err)
-		}
+			mux.Handle(callbackPath, oauthHandler)
+			redirectors[identityProvider.Usage.ProviderScope] = oauthHandler
+
+		case (*configapi.BasicAuthPasswordIdentityProvider),
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: use the IsPasswordAuthenticator() helper?

File: pkg/cmd/server/api/types.go [hunk 7]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+
+	// UseAsChallenger indicates whether to issue challenges for this provider
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: clarify `WWW-Authenticate challenges`

File: pkg/cmd/server/api/types.go [hunk 8]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+	UseAsChallenger bool
+	// UseAsLogin indicates whether to use this identity provider to back the login screen
+	UseAsLogin bool
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: does this have to be true for the oauth providers for them to redirect?

File: pkg/cmd/server/api/validation/validation.go [hunk 9]
```diff
@@ -86,6 +70,141 @@ func ValidateKubernetesMasterConfig(config *api.KubernetesMasterConfig) errs.Val
...
+	}
+
+	if config.SessionAuthenticationConfig != nil {
+		allErrs = append(allErrs, ValidateSessionAuthenticationConfig(config.SessionAuthenticationConfig).Prefix("sessionAuthenticationConfig")...)
+	}
+
+	allErrs = append(allErrs, ValidateGrantConfig(config.GrantConfig).Prefix("grantConfig")...)
+
+	for i, identityProvider := range config.IdentityProviders {
+		allErrs = append(allErrs, ValidateIdentityProvider(identityProvider).Prefix(fmt.Sprintf("identityProvider[%d]", i))...)
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: keep track of providers that set `UseAsLogin=true`, and fail if more than one does

Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: if a provider is a password provider and has `UseAsLogin=true`, require `config.SessionAuthenticationConfig` to be non-null

File: pkg/cmd/server/origin/auth.go [hunk 10]
```diff
@@ -322,230 +251,213 @@ func (c *AuthConfig) getAuthorizeAuthenticationHandlers(mux cmdutil.Mux) (authen
...
+				return nil, fmt.Errorf("unexpected error: %v", err)
+			}
 
-		state := external.DefaultState()
-		oauthHandler, err := external.NewExternalOAuthRedirector(oauthProvider, state, c.MasterPublicAddr+callbackPath, successHandler, errorHandler, identityMapper)
-		if err != nil {
-			return nil, fmt.Errorf("unexpected error: %v", err)
-		}
+			mux.Handle(callbackPath, oauthHandler)
+			redirectors[identityProvider.Usage.ProviderScope] = oauthHandler
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: only add here if `identityProvider.Usage.UseAsLogin` is set

File: pkg/cmd/server/origin/auth.go [hunk 11]
```diff
@@ -322,230 +251,213 @@ func (c *AuthConfig) getAuthorizeAuthenticationHandlers(mux cmdutil.Mux) (authen
...
+			}
+
+			if identityProvider.Usage.UseAsLogin {
+				redirectors["login"] = &redirector{RedirectURL: OpenShiftLoginPrefix, ThenParam: "then"}
+			}
+			if identityProvider.Usage.UseAsChallenger {
+				challengers["login"] = passwordchallenger.NewBasicAuthChallenger("openshift")
+			}
+
+			login := login.NewLogin(getCSRF(), &callbackPasswordAuthenticator{passwordAuth, successHandler}, login.DefaultLoginFormRenderer)
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: put this inside the `identityProvider.Usage.UseAsLogin` check as well

File: pkg/cmd/server/origin/auth.go [hunk 12]
```diff
@@ -322,230 +251,213 @@ func (c *AuthConfig) getAuthorizeAuthenticationHandlers(mux cmdutil.Mux) (authen
...
+	if c.getSessionAuth() != nil {
+		successHandlers = append(successHandlers, c.getSessionAuth())
 	}
 
-	switch c.AuthHandler {
-	case AuthHandlerGithub, AuthHandlerGoogle:
-		successHandlers = append(successHandlers, external.DefaultState().(handlers.AuthenticationSuccessHandler))
-	case AuthHandlerLogin:
-		successHandlers = append(successHandlers, redirectSuccessHandler{})
+	for _, identityProvider := range c.Options.IdentityProviders {
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: I think we have to check Usage.UseAsLogin first for all providers:
```
for ... {
  if !UseAsLogin {
    continue
  }
  if Oauth {
    DefaultState stuff
  }
  if Password {
    redirectSuccessHandler
  }
}
```

File: pkg/cmd/server/api/types.go [hunk 13]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+
+type DenyAllPasswordIdentityProvider struct {
+	api.TypeMeta
 }
 
+type HTPasswdPasswordIdentityProvider struct {
+	api.TypeMeta
+
+	// File is a reference to your htpasswd file
+	File string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: needs to make it into the GetMasterFileReferences list

File: pkg/cmd/server/api/types.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: needs to make it into the GetMasterFileReferences list

File: pkg/cmd/server/api/types.go [hunk 14]
```diff
@@ -146,9 +148,120 @@ type OAuthConfig struct {
...
+type XRemoteUserIdentityProvider struct {
+	api.TypeMeta
+
+	// CAFile is a file with the trusted signer certs
+	CAFile string
+	// Headers is the set of headers to check for identity information
+	Headers util.StringSet
+}
+
+type OAuthRedirectingIdentityProvider struct {
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: Not sure the deduplication of two fields is worth another nesting level. In reality, the google oauth provider will go away as soon as I can put together a generic openid identity provider

File: pkg/cmd/server/api/validation/validation.go [hunk 15]
```diff
@@ -86,6 +70,141 @@ func ValidateKubernetesMasterConfig(config *api.KubernetesMasterConfig) errs.Val
...
+			}
+
+		case (*api.BasicAuthPasswordIdentityProvider):
+			allErrs = append(allErrs, ValidateRemoteConnectionInfo(provider.RemoteConnectionInfo).Prefix("provider")...)
+
+		case (*api.HTPasswdPasswordIdentityProvider):
+			allErrs = append(allErrs, ValidateFile(provider.File, "provider.file")...)
+
+		case (*api.OAuthRedirectingIdentityProvider):
+			if len(provider.ClientID) == 0 {
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: make sure they don't try to use an OAuth provider as a challenger

File: pkg/cmd/server/api/validation/validation.go [hunk 16]
```diff
@@ -86,6 +70,141 @@ func ValidateKubernetesMasterConfig(config *api.KubernetesMasterConfig) errs.Val
...
+	if len(identityProvider.Usage.ProviderScope) == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired("usage.providerScope"))
+	}
+
+	if !api.IsIdentityProviderType(identityProvider.Provider) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("provider", identityProvider.Provider, fmt.Sprintf("%v is invalid in this context", identityProvider.Provider)))
+	} else {
+		switch provider := identityProvider.Provider.Object.(type) {
+		case (*api.XRemoteUserIdentityProvider):
+			if len(provider.CAFile) > 0 {
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: make sure they don't use a request header provider as a challenger or login

File: pkg/cmd/server/api/types.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: Marker for me to come back to this.  I realized last night that this is what was pooching my test.

File: pkg/cmd/server/api/types.go (hunk 14 above)
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: It's not just the deduplication of fields.  As a for instance, OAuth providers share validation restrictions.  It is possible to run all that with helpers, but I don't know that I'm a fan.  Construction of auth handlers for all types of oauth are also largely the same.

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
Comments addressed.

**Comment by liggitt** [comment]
My kingdom for polymorphism

**Comment by liggitt** [comment]
Ok, I'm pretty happy with this. @deads2k can you add examples of the different identity provider stanzas? @brenton, this would mean that if you want anything other than the default anypassword identity provider, you need to run from config. What's the best time/way for us to coordinate moving the RPM-installed version to run from config files?

**Comment by liggitt** [comment]
I also opened PR https://github.com/deads2k/origin/pull/8 against this branch to warn about deprecating the envvar method of configuring identity providers, and honor (for now) an envvar to deny login

**Comment by deads2k** [comment]
rebased.

**Comment by deads2k** [comment]
@sdodson @detiber This pull forces users to use a config file to use any identity provider other than "allow all".  It's a direction that makes sense for understandability of config and preventing mistakes, but it may affect the ansible config.  Any concerns before we merge?

File: pkg/cmd/server/api/types.go [hunk 17]
```diff
@@ -150,9 +149,120 @@ type OAuthConfig struct {
...
+type TokenConfig struct {
+	// Max age of authorize tokens
+	AuthorizeTokenMaxAgeSeconds int32
+	// Max age of access tokens
+	AccessTokenMaxAgeSeconds int32
+}
+
+type SessionConfig struct {
+	// SessionSecrets list the secret(s) to use to encrypt created sessions. Used by AuthRequestHandlerSession
+	SessionSecrets []string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: open a followup issue for me to make this structured. secrets have to be in pairs (one for signing, one for encryption, with specific length requirements on the encryption key)

**Comment by sdodson** [comment]
@deads2k This is user centric or do other parts use this? WRT to users, will `osc login` do the right thing so that users won't know that anything has changed?

**Comment by liggitt** [comment]
@sdodson no end-user impact, this is purely a cluster-admin config facing

**Comment by liggitt** [comment]
rebase, then LGTM

**Comment by deads2k** [comment]
[merge]

**Pull request merged**